### ISSUE TITLE
Segfault in 1.4.x istio-proxy with SD extension when starting up while receiving traffic

### ISSUE THREAD

**Issue by douglas-reid** [istio/istio#21505]
Setup:

- From a GCE VM (in this case: `ubuntu-1604-lts-drawfork-v20190930` running on `n1-standard-4 (4 vCPUs, 15 GB memory)`.
- ~15 QPS of incoming gRPC traffic (to the `productcatalogservice` from the hipster shop on port 3550)
- start envoy with metadata exchange filter and stackdriver filter (1.4.x versions)
- observe consistent Segfault.

Exemplar Stacktrace:

```
[Envoy (Epoch 0)] [2020-02-26 04:14:34.352][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:83] Caught Segmentation fault, suspect faulting address 0x20
[Envoy (Epoch 0)] [2020-02-26 04:14:34.352][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:70] Backtrace (use tools/stack_decode.py to get line numbers):
[Envoy (Epoch 0)] [2020-02-26 04:14:34.352][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71] Envoy version: 5f5d621c9fd0b68e12c8e1b95db849588fbdc475/1.12.0/Clean/RELEASE/BoringSSL
[Envoy (Epoch 0)] [2020-02-26 04:14:34.353][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #0: __restore_rt [0x7f07e3c45390]
[Envoy (Epoch 0)] [2020-02-26 04:14:34.361][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #1: Extensions::Stackdriver::Log::Logger::exportLogEntry() [0x5638dcf0b6a2]
[Envoy (Epoch 0)] [2020-02-26 04:14:34.369][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #2: Envoy::Extensions::Common::Wasm::Null::Plugin::Stackdriver::StackdriverRootContext::onTick() [0x5638dcf00a3d]
[Envoy (Epoch 0)] [2020-02-26 04:14:34.377][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #3: std::__1::__function::__func<>::operator()() [0x5638dd029467]
[Envoy (Epoch 0)] [2020-02-26 04:14:34.386][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #4: Envoy::Extensions::Common::Wasm::Wasm::tickHandler() [0x5638dd47337f]
[Envoy (Epoch 0)] [2020-02-26 04:14:34.394][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #5: std::__1::__function::__func<>::operator()() [0x5638dd47ee31]
[Envoy (Epoch 0)] [2020-02-26 04:14:34.402][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #6: event_process_active_single_queue [0x5638de299866]
[Envoy (Epoch 0)] [2020-02-26 04:14:34.410][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #7: event_base_loop [0x5638de29821e]
[Envoy (Epoch 0)] [2020-02-26 04:14:34.419][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #8: Envoy::Server::InstanceImpl::run() [0x5638ddf4fdd3]
[Envoy (Epoch 0)] [2020-02-26 04:14:34.427][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #9: Envoy::MainCommonBase::run() [0x5638dd0d2018]
[Envoy (Epoch 0)] [2020-02-26 04:14:34.435][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #10: main [0x5638dd0d0ff2]
[Envoy (Epoch 0)] [2020-02-26 04:14:34.435][4178][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #11: __libc_start_main [0x7f07e388a830]
```

If you turn off the incoming gRPC traffic, the Envoy starts and stays up. If you then turn the traffic back on, it continues to function properly without segfault.

Appears to be something around initialization.

I have not done extensive testing with 1.5, etc.

/cc @PiotrSikora

**Comment by PiotrSikora** [comment]
Just skimming through the extension's code, my guess is that once the traffic starts flowing, even if it's only internal (e.g. agent doing "are you up yet?" checks, or kubernetes health-checks), extension tries to send metrics/logs to Stackdriver, but Envoy didn't receive the CDS config yet, so it doesn't know about the target cluster and crashes instead of handling this gracefully.

Does this sound plausible, @bianpengyuan @jplevyak?

**Comment by mandarjog** [comment]
Saw this again with 1.4.x testing last night in a normal k8s setup.
However it was so pervasive that 3 out of 4 ingresses were crashlooping.

**Comment by bianpengyuan** [comment]
This could be related to excessive long start up time, which exaggerate the effect of this bug.

**Comment by mandarjog** [comment]
@bianpengyuan is https://github.com/istio/proxy/pull/2721. Fixed in 1.5 and master?
